### ISSUE TITLE
[Feature Request] API to explicitly set accessibility focus

### ISSUE THREAD

**Issue by luomo-pro** [flutter/flutter#94523]
First of all, thank the flutter team for their contribution to accessibility.
I am a blind person and a developer，Because I don't speak English, I use a translator, and the description may not be very accurate. If you have any questions, please ask.
There are many accessibility standards around the world, and one of them is a very important standard, which is to maintain a stable focus.
Currently,  flutter's management of accessibility focus is not perfect, and there are many problems.
For example, when you jump to a new page return, the Android side should set the accessibility focus to the previous position, but currently, it is set to the first control on the page.
The accessibility focus may be lost when the list item is rebuilt, and so on.
These problems can be solved using APIs that set accessibility focus.
Whether it is Android, iOS, web, or Windows, there are corresponding native APIs that can set the accessibility focus.

```
iOS：
UIAccessibilityPostNotification(UIAccessibilityLayoutChangedNotification, view);
```
```
Android:
view.sendAccessibilityEvent(AccessibilityEvent.TYPE_VIEW_FOCUSED);
```
```
web:
view.focus();
```
```
Windows:
NotifyWinEvent(EVENT_OBJECT_FOCUS, hwnd, OBJID_CLIENT, childId);
```

Now only the input focus can be set in flutter, but the accessibility focus cannot be set.
This issue was raised and discussed two years ago at the earliest, but it has not yet been resolved.
And I read the previous issue:
#91290
#59594
#77200

The result of these discussions is that although Android provides an API for setting accessibility focus, it is not intended to allow developers to actively set accessibility focus, and it may not take effect in the future.
But I want to say that in addition to Android, iOS, Windows, and the Web also have APIs for setting accessibility focus. Are they all redundant?
the people of Android said that it might not be allowed to take effect in the future, but it has been around for a long time, and there are many great accessibility applications, using this feature to meet the accessibility standards.
At least until now, APIs for setting accessibility focus on all platforms can work well.
Even if Android is no longer supported in the future, there are still iOS, Windows, web, so many platforms that support it.
That being the case, why doesn't flutter add an API for setting accessible focus now?

Some people say that if the developer actively sets the accessibility focus, it will make the application difficult to understand.
I want to say, is there any developer who will take the initiative to set the accessibility focus indiscriminately?
And flutter can now set the input focus, why has no one said to remove this API?
I believe that so many platforms have this API natively, and the purpose is definitely to make the software more accessible. Otherwise, why would they add this API?
Developers actively set accessibility focus to bring users a better experience.

Maybe because there is no blind person in the flutter team, so don't know the importance of this API to the disabled.
As a blind user and developer, I very much begged to increase the accessibility focus API!
It is very important to blind people all over the world!
And only in this way, a lot of software can be truly accessible, in line with global accessibility standards!
thank you very much!
@chunhtai

**Comment by chunhtai** [comment]
There are a lot of feature requests to add this functionality. The reason why we haven't added it is due to the Android team strongly discourage people from changing the accessibility focus. The Android API you mentioned work but it is an unofficial API, and it is not guaranteed to work. So it may not be maintainable if we put that in flutter. If this is very critical we can think about putting it in a third party package.

**Comment by chunhtai** [comment]
After some offline discussion with the team. I think it may make sense to fix the android behavior in flutter using the unofficial API. However, I still think we should not expose the  focus management API through flutter since it may not be maintainable.

@luomo-pro will this solve your use case?

**Comment by luomo-pro** [comment]
hi
@chunhtai
thank you very much!
I think that unless the system itself manages the accessibility focus very well, it would be very helpful to expose this API.
But up to now, there are still a lot of problems with the system itself in managing the accessibility focus. Even if it is native, there are some problems. The developer can only manage the accessibility focus manually.
In addition, in wcag (Web Content Accessibility Guidelines) accessibility standards, there is such a standard:
2.4.1 Bypass Blocks: A mechanism is available to bypass blocks of content that are repeated on multiple Web pages. (Level A) 
In many websites, developers will use the method of setting accessibility focus to meet this standard.
Therefore, I think it is necessary to set the accessibility focus API publicly now, so that developers can better comply with accessibility standards.
As for the Android team may no longer support this API in the future, I don't think you need to worry too much, there should be other ways to implement this feature at that time.
This API is already used by many softwares. As far as I know, there are China's Alibaba and Tencent. In order to ensure the stability of the accessibility focus in the software, they have manually set the accessibility focus.
In addition to Android, iOS, Windows, and Web also have this API. It can be seen that developers have a need for this API on all platforms.
Anyway, at least it works well now, isn't it?

**Comment by chunhtai** [comment]
My plan now would be try to better manage the focus in Flutter itself. I am still not confident enough to open up the API, the Android team strongly discouraging using the private API. They also mentioned such API may not be available in the future.

**Comment by wapscccs** [comment]
> [语义][Android]：开始新页面时焦点不返回页面顶部 #97522

Every time you open a new page, the accessibility focus will automatically fall on the back button, and a dialog box will pop up, and the accessibility focus will automatically fall on the title bar. Every time you return from a page to the previous interface, the focus will also fall automatically On the first element, these experiences are very bad for a blind user. I think you should simulate a blind person to operate the phone for a while and then look at this problem. You will find that this is really a very needed function. , otherwise I think the application developed with flutter is very bad for blind users. If there is a chance, I hope to communicate with you individually, thank you

**Comment by wapscccs** [comment]
> My plan now would be try to better manage the focus in Flutter itself. I am still not confident enough to open up the API, the Android team strongly discouraging using the private API. They also mentioned such API may not be available in the future.

Every time you open a new page, the accessibility focus will automatically fall on the back button, and a dialog box will pop up, and the accessibility focus will automatically fall on the title bar. Every time you return from a page to the previous interface, the focus will also fall automatically On the first element, these experiences are very bad for a blind user. I think you should simulate a blind person to operate the phone for a while and then look at this problem. You will find that this is really a very needed function. , otherwise I think the application developed with flutter is very bad for blind users. If there is a chance, I hope to communicate with you individually, thank you

**Comment by chunhtai** [comment]
@wapscccs 
Those issues you mentioned are what i am trying to fix, the idea is that Flutter will remember the previous focused item in a page and automatically shift the focus when that page comes to the front.

**Comment by luomo-pro** [comment]
I think these two things can be implemented simultaneously.
The first one is to allow flutter to manage accessibility focus better, including remembering exactly where the focus was on previous pages.
The second, the API for setting accessibility focus could be provided as a package, which would allow developers to manage it themselves in cases where flutter itself cannot manage it well, and could be done to complement it.
As I said before, with this API publicly available, developers don't use it indiscriminately, and the API is working fine now.

**Comment by ajanec01** [comment]
Yes, this issue is a pain! Is there a chance to work on a package that could circumvent it in the meantime @chunhtai ? At least for iOS. This would be one step forward.

**Comment by ajanec01** [comment]
Hi @chunhtai
Given that #97747 was moved to P2 a few weeks ago, is there a chance to change this issue to P2 as well?

Also, just thinking out laud now... if I were to implement native code through a method channel to try to manage VoiceOver focus whilst this issue is being looked into, do you you know how I can convert a Flutter Widget to a UIKit equivalent that I could pass as the `argument` to the native `.screenChanged` notification?

**Comment by chunhtai** [comment]
You would have to read through the UIAccessibility tree from the FlutterView, which is pretty hard I think. I am not sure if that is possible as I am not a objective c expert.

I think it would probably be easier to submit a PR to add the API to flutter. I am still not sure whether the API should support android, but It looks like we would need to support iOS.

**Comment by ajanec01** [comment]
Thanks @chunhtai !

The reason why I asked for this issue to be P2 is that in #97747 you commented:

> let's discuss the API for explicit control in the https://github.com/flutter/flutter/issues/94523. For this issue, we should focus on how or whether we should work around the android behaviour.

I thought that #97747 would sort out the way TalkBack focus behaves when pushing and popping routes but it would not try to resolve moving VoiceOver focus to a correct widget when you remove the widget that had the VO focus.

It would be great if that was added to the Flutter API, even with iOS support only to start with.

**Comment by chunhtai** [comment]
@dnfield If this is for iOS, then it is WAI as that is the native behavior, but like other had said, iOS let developer control the focus.

If this is for Android, they the other issue  https://github.com/flutter/flutter/issues/97747 may be the root cause.

> it would not try to resolve moving VoiceOver focus to a correct widget when you remove the widget that had the VO focus.

That is true, but we usually suggest to redesign to avoid that if possible. To a VO user, it would be confusing not matter which element the focus jump to. Compares to the Android problem where there isn't a workaround, thus the other issue has higher priority.

**Comment by dnfield** [comment]
(Issue I was referring to was for iOS. Developer specifically commented "we could control this in the native version of our app")

**Comment by chunhtai** [comment]
Yes, I think adding API to support platforms other than Android make sense.

@vick08 I remember we have talked about this before that we should not add API to set a11y focus for Android since it is not a stable API. Do we have guideline on other platforms like iOS, macOS, windows...etc?

**Comment by jacobsimionato** [comment]
A related issue came up in internal bug b/277986652 where an app is loading a new Android screen with a native bottom nave and Flutter top app bar, and focus is going to the bottom nav because Flutter loads asynchronously. Maybe exposing a focus changing API would make it easier for developers to fix this kind of integration issue, e.g. by focusing the Flutter view when it appears.

**Comment by tvolkert** [comment]
customer:chalk escalated this to me today as b/249445398, which is soon to be blocking them for iOS.  Updating priority to P2 here.

**Comment by hangyujin** [comment]
Current Plan:
Add a FocusSemanticEvent
Use the sendSemanticsEvent function in RenderObject
Handle the focus event on engine side

**Comment by hangyujin** [comment]
Update: working on prs https://github.com/flutter/flutter/pull/126171 and https://github.com/flutter/engine/pull/41777

**Comment by hangyujin** [comment]
PRs in review

**Comment by goderbauer** [comment]
Triage update: Looks like the engine change was just submitted, so we need to wait for a roll to continue on the framework side.

**Comment by chunhtai** [comment]
@hangyujin can this be closed?

**Comment by jjoelson** [comment]
@chunhtai Will this change address https://github.com/flutter/flutter/issues/97522, which was closed in favor of this issue? To clarify, the issue there was not returning focus to the original position after popping a page, but rather when pushing a new page the focus was not set to the top of the new page on Android.

**Comment by ajanec01** [comment]
Great stuff! Thank you!
Which version it's going to be in?
